raiden/messages/metadata.py

```diff
@@ -31,22 +30,22 @@ def _validate_address_metadata(self) -> None:
         for address in validation_errors:
             del self.address_metadata[address]
 
-    @cached_property
-    def hash(self) -> bytes:
-        return keccak(self._serialize_canonicaljson())
+    def _canonical_dict(self) -> dict:
+        """Return a dict that can be dumped as json
 
-    def _serialize_canonicaljson(self) -> bytes:
+        Used to build signatures.
+        """
         route = [to_checksum_address(address) for address in self.route]
-        if self.address_metadata is None:
-            address_metadata = None
-        else:
-            address_metadata = {
-                to_checksum_address(address): metadata
-                for address, metadata in self.address_metadata.items()
-            }
-        return canonicaljson.encode_canonical_json(
-            {"route": route, "address_metadata": address_metadata}
-        )
+        if not self.address_metadata:
+            # We add a default value of {} when validating. Normalize this to
+            # no key when signing.
+            return {"route": route}
+
+        address_metadata = {
+            to_checksum_address(address): metadata
+            for address, metadata in self.address_metadata.items()
+        }
+        return {"route": route, "address_metadata": address_metadata}
```

Comment:
Isn't this hardcoding the keys allowed on this schema? Can this support arbitrary unknown fields?

Reply:
The current implementation can't deal with unknown fields, yet. But we can add support for unknown fields without changing the encoding yet again. I wanted to keep this change as small as possible and add support for actually handling arbitrary data in a later PR.